Add reverse proxy mode

joshcooper opened the pull request:
Adds a reverse proxy mode whereby clients connect directly to the
reverse proxy, and the proxy forwards the connection to a target host
and port specified on the command line or config. For example:

```
$ python -m nogotofail.mitm -A invalidhostname \
  --mode reverse --port 443 --serverssl server.crt \
  --target_addr www.example.com --target_port 443
```

The reverse proxy mode makes it trivial to test a client that always
connects to a specific server port, as it doesn't require socks or
transparent proxy.

That said, the mode has some limitations. It cannot be used in cases
where the client needs to connect to multiple servers/ports. Also, it
doesn't rewrite HTTP headers, e.g. Host, so the values may be incorrect
when received by the real server, and when responses are received by the
client.

joshcooper committed:
Add reverse proxy mode

Adds a reverse proxy mode whereby clients connect directly to the
reverse proxy, and the proxy forwards the connection to a target host
and port specified on the command line or config. For example:

    $ python -m nogotofail.mitm -A invalidhostname \
      --mode reverse --port 443 --serverssl server.crt \
      --target_addr www.example.com --target_port 443

The reverse proxy mode makes it trivial to test a client that always
connects to a specific server port, as it doesn't require socks or
transparent proxy.

That said, the mode has some limitations. It cannot be used in cases
where the client needs to connect to multiple servers/ports. Also, it
doesn't rewrite HTTP headers, e.g. Host, so the values may be incorrect
when received by the real server, and when responses are received by the
client.

chadbrubaker commented on nogotofail/mitm/__main__.py:
```diff
@@ -127,6 +127,10 @@ def set_tproxy_rules(args):
     routing.enable_tproxy_rules(port, ipv6=ipv6)
     atexit.register(routing.disable_tproxy_rules, ipv6=ipv6)
 
+def set_reverse_rules(args):
```
Can this explicitly check if target_addr and target_port were set and fail with an obvious error if not?

joshcooper replied:
Yes makes sense. I'll update the PR.

joshcooper replied:
I added validation. I'm not sure if using `ValueError` is best practice?

```
$ python -m nogotofail.mitm --mode reverse
2016-05-13 11:45:39,498 [INFO] Starting...
Error: Target address is required
$ python -m nogotofail.mitm --mode reverse --target_addr foo
2016-05-13 11:45:45,929 [INFO] Starting...
Error: Target port must be between 1 and 65535
$ python -m nogotofail.mitm --mode reverse --target_addr foo --target_port 0
2016-05-13 11:45:56,466 [INFO] Starting...
Error: Target port must be between 1 and 65535
```

joshcooper committed:
Validate target addr and port

Target addr and port are required when using reverse mode. Print an
error to stderr if they are omitted or incorrect and exit with 2,
similar to what happens when the type of argument doesn't match.

joshcooper commented:
Hi, I was wondering if there's anything I can do to help move this forward? I assume using `ValueError` is sufficient?

chadbrubaker commented:
Sorry for the delay

joshcooper commented:
No worries, thanks @chadbrubaker!
